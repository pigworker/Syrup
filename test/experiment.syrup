not(<Bit>) -> <Bit>
not(X) = nand(X,X)

and(<Bit>, <Bit>) -> <Bit>
and(X,Y) = not(nand(X,Y))

or(<Bit>, <Bit>) -> <Bit>
or(X,Y) = nand(Z, S) where
  Z = not(X)
  S = not(Y)

xor(<Bit>, <Bit>) -> <Bit>
xor(X,Y) = or(L,R) where

  L = and(X,not(Y))
  R = and(not(X),Y)

hadd(<Bit>, <Bit>) -> <Bit>, <Bit>
hadd(X, Y) = X&Y, xor(X,Y)

experiment hadd

fadd(<Bit>, <Bit>, <Bit>) -> <Bit>, <Bit>
fadd(X, Y, CI) = CO, Z where

  C1, XY = hadd(X,Y)
  C2, Z  = hadd(XY,CI)
  CO = C1|C2

experiment fadd

type <Two> = [<Bit>, <Bit>]

adc2(<Two>, <Two>, <Bit>) -> <Bit>, <Two>
adc2([X1, X0], [Y1, Y0], CI) = CO, [Z1, Z0] where

  C1, Z0 = fadd(X0, Y0, CI)
  CO, Z1 = fadd(X1, Y1, C1)

experiment adc2

experiment adc2([10][01] 1; [11][10] 0)

xor1(<Bit>, <Bit>) -> <Bit>
xor1(X,Y) = !X&Y | X&!Y

xor2(<Bit>, <Bit>) -> <Bit>
xor2(X,Y) = (X|Y) & !(X&Y)

experiment xor1 = xor2

experiment xor = or

print adc2