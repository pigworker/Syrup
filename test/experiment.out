not is defined.

and is defined.

or is defined.

xor is defined.

hadd is defined.

Truth table for hadd:
X Y |    
----|----
0 0 | 0 0
0 1 | 0 1
1 0 | 0 1
1 1 | 1 0

fadd is defined.

Truth table for fadd:
X Y CI | CO Z
-------|-----
0 0 0  | 0  0
0 0 1  | 0  1
0 1 0  | 0  1
0 1 1  | 1  0
1 0 0  | 0  1
1 0 1  | 1  0
1 1 0  | 1  0
1 1 1  | 1  1

Type alias Two defined.

adc2 is defined.

Truth table for adc2:
[X1 X0] [Y1 Y0] CI | CO [Z1 Z0]
-------------------|-----------
[0  0 ] [0  0 ] 0  | 0  [0  0 ]
[0  0 ] [0  0 ] 1  | 0  [0  1 ]
[0  0 ] [0  1 ] 0  | 0  [0  1 ]
[0  0 ] [0  1 ] 1  | 0  [1  0 ]
[0  0 ] [1  0 ] 0  | 0  [1  0 ]
[0  0 ] [1  0 ] 1  | 0  [1  1 ]
[0  0 ] [1  1 ] 0  | 0  [1  1 ]
[0  0 ] [1  1 ] 1  | 1  [0  0 ]
[0  1 ] [0  0 ] 0  | 0  [0  1 ]
[0  1 ] [0  0 ] 1  | 0  [1  0 ]
[0  1 ] [0  1 ] 0  | 0  [1  0 ]
[0  1 ] [0  1 ] 1  | 0  [1  1 ]
[0  1 ] [1  0 ] 0  | 0  [1  1 ]
[0  1 ] [1  0 ] 1  | 1  [0  0 ]
[0  1 ] [1  1 ] 0  | 1  [0  0 ]
[0  1 ] [1  1 ] 1  | 1  [0  1 ]
[1  0 ] [0  0 ] 0  | 0  [1  0 ]
[1  0 ] [0  0 ] 1  | 0  [1  1 ]
[1  0 ] [0  1 ] 0  | 0  [1  1 ]
[1  0 ] [0  1 ] 1  | 1  [0  0 ]
[1  0 ] [1  0 ] 0  | 1  [0  0 ]
[1  0 ] [1  0 ] 1  | 1  [0  1 ]
[1  0 ] [1  1 ] 0  | 1  [0  1 ]
[1  0 ] [1  1 ] 1  | 1  [1  0 ]
[1  1 ] [0  0 ] 0  | 0  [1  1 ]
[1  1 ] [0  0 ] 1  | 1  [0  0 ]
[1  1 ] [0  1 ] 0  | 1  [0  0 ]
[1  1 ] [0  1 ] 1  | 1  [0  1 ]
[1  1 ] [1  0 ] 0  | 1  [0  1 ]
[1  1 ] [1  0 ] 1  | 1  [1  0 ]
[1  1 ] [1  1 ] 0  | 1  [1  0 ]
[1  1 ] [1  1 ] 1  | 1  [1  1 ]

Simulation for adc2:
0 {} [10][01]1 -> 1[00]
1 {} [11][10]0 -> 1[01]
2 {}

xor1 is defined.

xor2 is defined.

xor1 behaves like xor2
  {} ~ {}

xor has a behaviour that or does not match
  xor(11) = 0 but or(11) = 1

adc2([<Bit>, <Bit>], [<Bit>, <Bit>], <Bit>) -> <Bit>, [<Bit>, <Bit>]

adc2([<Bit>, <Bit>], [<Bit>, <Bit>], <Bit>) -> <Bit>, [<Bit>, <Bit>]
adc2([X1, X0], [Y1, Y0], CI) = CO, [Z1, Z0] where
  C1, Z0 = fadd(X0, Y0, CI)
  CO, Z1 = fadd(X1, Y1, C1)


